File: mixin-base/src/main/kotlin/net/yakclient/mixin/base/internal/loader/ContextPool.java
package net.yakclient.mixin.base.internal.loader;

import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.Nullable;

import java.util.HashMap;
import java.util.Map;

public abstract class ContextPool {
    private final Map<PackageTarget, Context> targets;

    private final ClassLoader loader;

    public ContextPool() {
        this.targets = new HashMap<>();
        this.loader = ClassLoader.getSystemClassLoader();
//        ContextPoolManager.
    }

    private boolean encapsulatedTargetExists(PackageTarget target) {
        for (PackageTarget packageTarget : this.targets.keySet()) {
            if (packageTarget.isTargetOf(target)) return true;
        }
        return false;
    }

    public PackageTarget encapsulatedTarget(PackageTarget target) {
        if (!this.encapsulatedTargetExists(target)) throw new IllegalStateException("Target MUST be encapsulated!");
        for (PackageTarget packageTarget : this.targets.keySet()) {
            if (packageTarget.isTargetOf(target)) return packageTarget;
        }
        throw new IllegalStateException("Target MUST be encapsulated!");
    }

    public Context addTarget(PackageTarget target, ClassLoader loader) {
        final PackageTarget realTarget = this.getTarget(target);
        return this.targets.put(realTarget, this.createContext(loader, realTarget));
    }

    public Context addTarget(PackageTarget target) {
        final PackageTarget realTarget = this.getTarget(target);
        final ClassLoader loader = this.targets.containsKey(realTarget) ? this.targets.get(realTarget).getLoader() : ContextPoolManager.createLoader(realTarget);

        final Context context = this.createContext(loader, target);
        this.targets.putIfAbsent(this.getTarget(target), context);
        return context;
    }

    @Contract(pure = true)
    public PackageTarget getTarget(PackageTarget target) {
        return this.encapsulatedTargetExists(target) ? this.encapsulatedTarget(target) : target;
    }

    public boolean isTargeted(String target) {
        final PackageTarget packageTarget = this.createTarget(target);
        return this.isTargeted(packageTarget);
    }

    public boolean isTargeted(PackageTarget target) {
        for (final PackageTarget pT : this.targets.keySet()) {
            if (pT.isTargetOf(target)) return true;
        }
        return false;
    }

    @Nullable
    public Class<?> loadClassOrNull(String name) {
        final PackageTarget target = this.getTarget(this.createTarget(name));
        return this.targets.get(target).findClass(name);
    }

    private Context createContext(ClassLoader loader, PackageTarget target) {
        return new Context(loader, target, this);
    }

    PackageTarget createTarget(String path) {
        try {
            if (this.loader.loadClass(path) != null) return ClassTarget.of(path);
            return PackageTarget.of(path);
        } catch (ClassNotFoundException e) {
            return PackageTarget.of(path);
        }
    }

    public Class<?> defineClass(ClassTarget target, byte[] bytes) {
        final PackageTarget finalTarget = this.getTarget(target);
        final Context context = this.targets.get(finalTarget);
        final String name = target.toString();

        if (!context.getLoader().isDefined(name)) return context.getLoader().defineClass(name, bytes);
        else
            this.targets.put(finalTarget, this.createContext(ContextPoolManager.createLoader(finalTarget), finalTarget));

        return this.targets.get(finalTarget).getLoader().defineClass(name, bytes);
    }
}

File: mixin-base/src/main/kotlin/net/yakclient/mixin/base/internal/loader/ContextPool.kt
package net.yakclient.mixin.base.internal.loader

import net.yakclient.mixin.base.internal.loader.ContextPoolManager.createLoader
import net.yakclient.mixin.base.target.ClassTarget
import net.yakclient.mixin.base.target.PackageTarget
import net.yakclient.mixin.base.target.Target
import org.jetbrains.annotations.Contract

abstract class ContextPool {
    private val targets: MutableMap<Target, Context>
    private val loader: ClassLoader

    init {
        targets = HashMap()
        loader = ClassLoader.getSystemClassLoader()
    }

    private fun safelyEncapsulate(target: Target): Target {
        for (encapsulated in targets.keys) {
            if (encapsulated.isTargetOf(target)) return encapsulated
        }
        return target
    }

    fun addTarget(target: Target): Context {
        val realTarget = safelyEncapsulate(target)
        val loader: ProxyClassLoader =
            if (targets.containsKey(realTarget)) targets[realTarget]!!.loader else createLoader(realTarget)
        val context = createContext(loader, target)
        if (!targets.containsKey(realTarget)) targets[realTarget] = context
        return context
    }

    fun isTargeted(target: Target): Boolean {
        for (pT in targets.keys) {
            if (pT.isTargetOf(target)) return true
        }
        return false
    }

    fun loadClassOrNull(target: Target, name: String): Class<*>? {
        val context = checkNotNull(targets[target]) { "Failed to find context from given target: ${target.name()}" }
        return context.findClass(name)
    }

    fun isClassDefined(target: Target, name: String): Boolean {
        val context = checkNotNull(targets[target]) { "Failed to find context from given target: ${target.name()}" }
        return context.loader.isDefined(name)
    }

    fun defineClass(target: Target, name: String, bytes: ByteArray): Class<*> {
        val finalTarget = safelyEncapsulate(target)
        val oldContext = checkNotNull(targets[finalTarget]) { "Failed to find appropriate context to define class in" }

        val context = if (oldContext.loader.isDefined(name)) this.setContext(
            finalTarget,
            createContext(createLoader(finalTarget), finalTarget)
        ) else oldContext

        return context.loader.defineClass(name, bytes)
    }

    private fun setContext(target: Target, context: Context): Context {
        targets[target] = context
        return context
    }

    private fun createContext(loader: ProxyClassLoader, target: Target): Context {
        return Context(loader, target, this)
    }
}